.teamcity/_self/lib/customBuildType/E2EBuildType.kt

```diff
@@ -0,0 +1,170 @@
+ * @param buildParams Environment variables and other parameters to set for the build.
+ * @param buildFeatures Features to enable on top of the default feature set (perfmon, commitStatusPublisher).
+ * @param buildTriggers Rules to trigger the build. By default, no triggers are defined.
+ * @param buildDepedencies If the build configuration depends on another existing build configuration, define it here.
+ */
+open class E2EBuildType(
+	var buildId: String,
+	var buildUuid: String,
+	var buildName: String,
+	var buildDescription: String,
+	var concurrentBuilds: Int = 0,
+	var getCalypsoLiveURL: String = "",
+	var testGroup: String,
+	var buildParams: ParametrizedWithType.() -> Unit = {},
+	var buildFeatures: BuildFeatures.() -> Unit,
+	var buildTriggers: Triggers.() -> Unit = {},
+	var buildDependencies: Dependencies.() -> Unit = {},
+
+): BuildType() {
+	init {
+		val concurrentBuilds = concurrentBuilds
+		val getCalypsoLiveURL = getCalypsoLiveURL
+		val testGroup = testGroup
+		val buildParams = buildParams
+		val buildFeatures = buildFeatures
+		val buildTriggers = buildTriggers
+		val buildDependencies = buildDependencies
+		val params = params
+
+		id( buildId )
+		uuid = buildUuid
+		name = buildName
+		description = buildDescription
+		maxRunningBuilds = concurrentBuilds
+
+		artifactRules = """
+			logs.tgz => logs.tgz
+			screenshots => screenshots
+			trace => trace
+		""".trimIndent()
+
+		vcs {
+			root(Settings.WpCalypso)
+			cleanCheckout = true
+		}
+
+		params {
+			param("env.NODE_CONFIG_ENV", "test")
+			param("env.PLAYWRIGHT_BROWSERS_PATH", "0")
+			param("env.TEAMCITY_VERSION", "2021")
+			param("env.HEADLESS", "false")
+			param("env.LOCALE", "en")
+			param("env.DEBUG", "pw:api")
+			buildParams()
+		}
+
+		steps {
+			prepareEnvironment()
+
+			bashNodeScript {
+				name = "Run tests"
+				scriptContent = """
+					# Configure bash shell.
+					shopt -s globstar
+					set -x
+
+					# For Calypso E2E build configurations, the URL environment variable
+					# is computed and exported by a script.
+					$getCalypsoLiveURL
+
+					# Enter testing directory.
+					cd test/e2e
+					mkdir temp
+
+					# Decrypt config
+					openssl aes-256-cbc -md sha1 -d -in ./config/encrypted.enc -out ./config/local-test.json -k "%E2E_CONFIG_ENCRYPTION_KEY%"
+
+					# As noted above, Calypso E2E build configurations export the URL
+					# environment variable within this step - therefore the export
+					# for NODE_CONFIG variable has to be done here instead of
+					# within the Kotlin DSL.
+					export NODE_CONFIG="{\"calypsoBaseURL\":\"${'$'}{URL%/}\"}"
+
+					# Run suite.
+					xvfb-run yarn jest --reporters=jest-teamcity --reporters=default --maxWorkers=%E2E_WORKERS% --group=$testGroup
+				"""
+				dockerImage = "%docker_image_e2e%"
+			}
+
+			collectResults()
+		}
+
+		features {
+			perfmon {
+			}
+			commitStatusPublisher {
+				vcsRootExtId = "${Settings.WpCalypso.id}"
```

Comment:
I think so too - this was just copy/pasted over from the original config. Thanks!

Reply:
I tried removing the string part, but it seems it is required, otherwise a compilation error results:

```
Compilation error _self/lib/customBuildType/E2EBuildType.kt[156:20]: Type mismatch: inferred type is Id? but String? was expected
```

Reply:
Unfortunate!